### PR TITLE
Feature/project release management

### PR DESCRIPTION
**Pull request by hujiahao-hjh**
## What this PR does / why we need it:
Add project release management.

## I have checked the following points:
- [x] I18n is finished and updated by cli
- [x] Form fields validation is added and length is limited
- [x] Display normally on small screen
- [x] Display normally when some data is empty or null
- [x] Display normally in english mode

## Does this PR introduce a user interface change?

❎ No

## ChangeLog

| Language | Changelog |
| --------- | ------------ |
| 🇺🇸 English | Added project level release management |
| 🇨🇳 中文    |  新增项目级别的制品管理。 |

## Does this PR need be patched to older version?
❎ No

## Which issue(s) this PR fixes:

Fixes #

### CODE REVIEW

File: shell/app/common/components/list-select/index.scss [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+
+    &::-moz-placeholder {
+      /* Mozilla Firefox 19+ */
+      color: $color-white-3;
+    }
+
+    &:-ms-input-placeholder {
+      /* Internet Explorer 10+ */
+      color: $color-white-3;
+    }
```
Comment (daskyrk) [inline shell/app/common/components/list-select/index.scss:167]: Not support for IE anymore.

File: shell/app/common/components/list-select/index.tsx [hunk 2]
```diff
@@ -0,0 +1,343 @@
...
+      (index === 0 || index) && currentList.splice(index, 1);
+    }
+    setSelectedList(currentList);
+  };
+
+  const remove = (id: string) => {
+    const currentList = [...selectedList];
+    const index = selectedList.findIndex((item) => item[rowKey] === id);
+    currentList.splice(index, 1);
+    setSelectedList(currentList);
```
Comment (daskyrk) [inline shell/app/common/components/list-select/index.tsx:59]: Just `setSelectedList(prev => prev.filter(item => ...))` is ok

File: locales/en.json [hunk 3]
```diff
@@ -641,6 +641,7 @@
     "ok": "ok",
     "on": "on",
     "one key all read": "one key all read",
+    "on key to empty": "on key to empty",
```
Comment (McDaddy) [inline locales/en.json:644]: on key ?

File: shell/app/common/components/render-form-item/index.tsx [hunk 4]
```diff
@@ -291,9 +297,35 @@ const RenderFormItem = ({
...
+        ItemComp = readOnlyRender ? (
+          <CustomRender readOnlyRender={readOnlyRender} />
+        ) : (
+          <ClassWrapper>
+            <MarkdownReadOnlyRender {...itemProps} />
+          </ClassWrapper>
+        );
+      }
+      break;
+    }
```
Comment (daskyrk) [inline shell/app/common/components/render-form-item/index.tsx:322]: Not include custom component here, use custom type

File: locales/en.json (hunk 3 above)
Comment (McDaddy) [reply locales/en.json:644]: one click to clear is better

File: shell/app/config-page/components/table/render-types.tsx [hunk 5]
```diff
@@ -630,6 +630,7 @@ const getTableOperation = (val: any, record: any, extra: any) => {
             }}
             key={key}
             onCancel={(e: any) => e && e.stopPropagation()}
+            zIndex={1100}
```
Comment (daskyrk) [inline shell/app/config-page/components/table/render-types.tsx:633]: Is this useful ?

File: shell/app/modules/project/pages/release/application.tsx [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+import ReleaseProtocol from './release-protocol';
+import releaseStore from 'project/stores/release';
+import routeInfoStore from 'core/stores/route';
+import { useLoading } from 'core/stores/loading';
+
+import './index.scss';
+
+const ProjectRelease = () => {
+  const { params } = routeInfoStore.getState((s) => s);
+  const { appList } = releaseStore.getState((s) => s);
```
Comment (daskyrk) [inline shell/app/modules/project/pages/release/application.tsx:26]: Not use `s => s`, any change in the store will trigger render.

Comment (daskyrk) [reply shell/app/modules/project/pages/release/application.tsx:26]: Why use getState but not useStore ?

File: shell/app/common/components/list-select/index.scss [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+      /* WebKit browsers */
+      color: $color-white-3;
+    }
+
+    &::-moz-placeholder {
+      /* Mozilla Firefox 19+ */
+      color: $color-white-3;
+    }
+
+    &:-ms-input-placeholder {
```
Comment (McDaddy) [inline shell/app/common/components/list-select/index.scss:164]: dont just take care the place where be commented

File: shell/app/common/components/list-select/index.tsx [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+
+  React.useEffect(() => {
+    onMenuChange && onMenuChange(menus.find((item) => item[menuRowKey] === selectedMenu));
+  }, [selectedMenu, menus, onMenuChange, menuRowKey]);
+
+  return (
+    <Row className="erda-list-select-overlay text-white rounded">
+      <Col span={12} className="px-2 h-full">
+        <div className="py-3 px-2">
+          {i18n.t('selected {xx}', { xx: label })}
```
Comment (McDaddy) [inline shell/app/common/components/list-select/index.tsx:202]: dont be some casual

File: shell/app/common/components/list-select/index.tsx [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+import { Dropdown, Row, Col, Button, Input } from 'antd';
+import i18n from 'i18n';
+import { ErdaIcon, Pagination } from 'common';
+
+import './index.scss';
+
+interface IProps extends ListSelectOverlayProps {
+  label: React.ReactNode;
+  renderSelectedItem?: (item: Obj) => React.ReactNode;
+  value: Obj[];
```
Comment (McDaddy) [inline shell/app/common/components/list-select/index.tsx:24]: can this `Obj` have generic？

File: shell/app/modules/project/pages/release/components/application-detail.tsx [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+import { useLoading } from 'core/stores/loading';
+
+import './form.scss';
+
+const { TabPane } = Tabs;
+
+const ReleaseApplicationDetail = () => {
+  const { params } = routeInfoStore.getState((s) => s);
+  const { releaseID } = params;
+  const { releaseDetail } = releaseStore.getState((s) => s);
```
Comment (daskyrk) [inline shell/app/modules/project/pages/release/components/application-detail.tsx:36]: Same as above

File: shell/app/modules/project/stores/release.ts [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+      const res = await call(getAppList, payload);
+      const { list } = res;
+      update({ appList: list.map((item) => ({ ...item, title: item.displayName })) });
+    },
+    async getReleaseList({ call, update }, payload: RELEASE.ReleaseListQuery) {
+      const res = await call(getReleaseList, payload);
+      const { list, total } = res;
+      update({ releaseList: list, releaseTotal: total });
+    },
+    async addRelease({ call, getParams }, payload): Promise<{ success: boolean }> {
```
Comment (daskyrk) [inline shell/app/modules/project/stores/release.ts:57]: Missing payload type

File: shell/app/common/components/list-select/index.tsx [hunk 12]
```diff
@@ -0,0 +1,340 @@
...
+  const [selectedList, setSelectedList] = React.useState<Obj[]>([]);
+  const [resultList, setResultList] = React.useState<Obj[]>([]);
+  const [visible, setVisible] = React.useState<boolean>(false);
+  const select = (selectItem: Obj, checked: boolean) => {
+    const currentList = [...selectedList];
+    if (checked) {
+      currentList.push(selectItem);
+    } else {
+      const index = currentList.findIndex((item) => item[rowKey] === selectItem[rowKey]);
+      (index === 0 || index) && currentList.splice(index, 1);
```
Comment (McDaddy) [inline shell/app/common/components/list-select/index.tsx:50]: use filter instead

File: shell/app/common/components/list-select/index.tsx [hunk 13]
```diff
@@ -0,0 +1,340 @@
...
+    setResultList(currentList);
+  };
+
+  const clear = () => {
+    setSelectedList([]);
+  };
+
+  const onOk = () => {
+    onChange(selectedList);
+    setResultList([...selectedList]);
```
Comment (McDaddy) [inline shell/app/common/components/list-select/index.tsx:73]: why need deconstruction？

File: shell/app/common/components/list-select/index.tsx [hunk 14]
```diff
@@ -0,0 +1,340 @@
...
+  clear: () => void;
+  selectedList: Obj[];
+  rowKey: string;
+  menuRowKey: string;
+  parentKey: string;
+  value: Obj[];
+  onChange: (values: Obj[]) => void;
+}
+
+interface IPagination {
```
Comment (McDaddy) [inline shell/app/common/components/list-select/index.tsx:156]: what's the difference between other `IPagination`? maybe u can extend from antd props

File: shell/app/common/components/list-select/index.tsx [hunk 15]
```diff
@@ -0,0 +1,340 @@
...
+  current: number;
+  pageSize: number;
+  total: number;
+  onChange: (pageNo: number, pageSize: number) => void;
+}
+
+const defaultRenderItem = (item: Obj) => {
```
Comment (McDaddy) [inline shell/app/common/components/list-select/index.tsx:163]: Keep Obj with generic

File: shell/app/common/components/list-select/index.tsx [hunk 16]
```diff
@@ -0,0 +1,340 @@
...
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+import React from 'react';
+import { Dropdown, Row, Col, Button, Input } from 'antd';
+import i18n from 'i18n';
+import { ErdaIcon, Pagination } from 'common';
+
+import './index.scss';
+
+interface IProps extends ListSelectOverlayProps {
```
Comment (McDaddy) [inline shell/app/common/components/list-select/index.tsx:21]: if its top level common component, docs is required

File: shell/app/common/components/list-select/index.tsx [hunk 17]
```diff
@@ -0,0 +1,340 @@
...
+  label: React.ReactNode;
```
Comment (McDaddy) [inline shell/app/common/components/list-select/index.tsx:22]: Also where is the unit test case

File: shell/app/modules/project/pages/release/components/form.tsx [hunk 18]
```diff
@@ -0,0 +1,236 @@
...
+          releaseID,
+        }).then((res: { success: boolean }) => {
+          if (res.success) {
+            goTo(goTo.pages.projectRelease);
+          }
+        });
+      } else {
+        addRelease({
+          ...payload,
+        }).then((res: { success: boolean }) => {
```
Comment (McDaddy) [inline shell/app/modules/project/pages/release/components/form.tsx:210]: RAW_RESPONSE instead

File: shell/app/modules/project/pages/release/components/project-detail.tsx [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+          if (res.success) {
+            getReleaseDetail({ releaseID: releaseId });
+          }
+        });
+      },
+    });
+  };
+
+  return (
+    <div className="release-detail">
```
Comment (McDaddy) [inline shell/app/modules/project/pages/release/components/project-detail.tsx:46]: you are reusing the style from another component?

File: shell/app/modules/project/pages/release/components/project-detail.tsx [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+
+import React from 'react';
+import { Button, Spin, Modal } from 'antd';
+import ReleaseForm from './form';
+import i18n from 'i18n';
+import { goTo } from 'common/utils';
+import releaseStore from 'project/stores/release';
+import { useLoading } from 'core/stores/loading';
+
+import './form.scss';
```
Comment (McDaddy) [inline shell/app/modules/project/pages/release/components/project-detail.tsx:22]: I feel very strange, why import form.scss here

File: shell/app/modules/project/pages/release/release-protocol.tsx [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+  const [{ projectId }] = routeInfoStore.useStore((s) => [s.params]);
+  const [isFormal, setIsFormal] = React.useState<string>('informal');
+  const inParams = {
+    isProjectRelease,
+    isFormal: isFormal === 'formal',
+    projectID: +projectId,
+    applicationID,
+  };
+
+  const reloadRef = React.useRef(null as any);
```
Comment (McDaddy) [inline shell/app/modules/project/pages/release/release-protocol.tsx:37]: remove any

File: shell/app/modules/project/pages/release/release-protocol.tsx [hunk 22]
```diff
@@ -0,0 +1,92 @@
...
+            {i18n.t('new {name}', { name: i18n.t('Artifact') })}
+          </Button>
+        </div>
+      ) : null}
+
+      <RadioTabs
+        options={options}
+        value={isFormal}
+        onChange={(v?: string | number) => {
+          setIsFormal(v as string);
```
Comment (McDaddy) [inline shell/app/modules/project/pages/release/release-protocol.tsx:66]: in line 65 you defined v as string | number and optional. In the next line, you use `as` to type it as string. I dont understand

File: shell/app/modules/project/pages/release/index.scss [hunk 23]
```diff
@@ -0,0 +1,37 @@
+.release-app-list {
```
Comment (McDaddy) [inline shell/app/modules/project/pages/release/index.scss:1]: naming this as index.scss. but I don't find index.tsx

File: shell/app/modules/project/types/release.d.ts [hunk 24]
```diff
@@ -0,0 +1,55 @@
...
+    projectId?: string;
+    q?: string;
+  }
+
+  interface AppDetail {
+    id: number;
+    displayName: string;
+  }
+
+  interface ReleaseListQuery {
```
Comment (McDaddy) [inline shell/app/modules/project/types/release.d.ts:25]: define types like query in service or store

File: shell/app/modules/project/types/release.d.ts [hunk 25]
```diff
@@ -0,0 +1,70 @@
...
+    applicationId?: string | number;
+    pageSize: number;
+    pageNo: number;
+    q?: string;
+  }
+
+  interface AddReleaseParams extends ReleaseDetail {
+    applicationReleaseList: string[];
+    isStable: boolean;
+    isFormal: boolean;
```
Comment (McDaddy) [inline shell/app/modules/project/types/release.d.ts:39]: id is optional in detail?

**Pull request merged**